### PR TITLE
revamp crucible return code handling and error logging

### PR DESCRIPTION
**Pull request by k-rister**
- functions explicitly return values as opposed to setting a global
  variable making it easier to determine the who/what/where/when of a
  return code change

- in a lot of places return codes were being ignored so fix that.
  this required some restructing of code in places where there was no
  hierarchy to conditionally handle the pass/fail status of a previous
  command.

- make sure ${PIPESTATUS[]} is used to properly evaluate the return
  code of the "interesting" command(s)

### CODE REVIEW

**Comment by k-rister** [comment]
@rafaelfolco @atheurer
I'm not convinced that all of my code change decisions are the best choices.  In some of the places that needed updating there was so little hierarchy that I had to come up with something and it might not be the best.  Feel free to make suggestions of better approaches.

**Comment by k-rister** [comment]
As I expected crucible-ci failed because crucible is broken following a recent rickshaw update.  Without the fixes in this PR the rickshaw bugs were allowed to pass crucible-ci test and be merged.

We either need to get rickshaw fixed and crucible-ci passing again before merging this, or we will have to disable the the status check in order to merge this -- once we agree that the logic is sound.

File: bin/_main
Comment (rafaelfolco) [inline bin/_main:216]: return code 1 may be convenient for the time being, however it might be a problem to test and get the expected result... You may need to grep for the stdout to see if your code is behaving properly....

**Review by rafaelfolco — APPROVED** [review]
as far as I can tell, this looks good

**Comment by k-rister** [comment]
@rafaelfolco
I just pushed a big update...

File: bin/_main
Comment (k-rister) [reply bin/_main]: ack

File: bin/_main
Comment (k-rister) [reply bin/_main]: ack

File: bin/_main
Comment (k-rister) [reply bin/_main]: ack

**Review by rafaelfolco — APPROVED** [review]
lgtm

**Pull request merged**
